Chess/state.h

```diff
@@ -0,0 +1,87 @@
+#ifndef STATE_H
+#define STATE_H
+
+#include "common_types.h"
+#include "game.h"
+
+class Game;
+
+class State
+{
+public:
+    State();
+    virtual ~State();
+
+    virtual void cellAction(Game*, QString);
+
+    virtual void startAction(Game*);
+    virtual void stopAction(Game*);
+
+    virtual void saveAction(Game*, QString);
+    virtual bool loadAction(Game*, QString);
+
+    virtual void rollback_from_list(Game*);
+    virtual void make_move_from_list(Game*);
+
+protected:
+    void ChangeState (Game*, State*);
+
+};
+
+
+/////////////////////////////////////////////////////////////////////////
+class State1 : public State
+{
+public:
+    State1() {}
+
+    static State* Instance();
+
+    void cellAction(Game*, QString);
+    void startAction(Game*);
+    bool loadAction(Game*, QString);
+
+private:
+    static State *self_;
+
+};
+
+/////////////////////////////////////////////////////////////////////////
+class State2 : public State
+{
+public:
+    State2() {}
+
+    static State* Instance();
```

Comment:
I have met some troubles with Mayers singlton and derived classes, so use this implementation as was suggested in this topic (C++ implementation block)
https://ru.wikipedia.org/wiki/%D0%9E%D0%B4%D0%B8%D0%BD%D0%BE%D1%87%D0%BA%D0%B0_(%D1%88%D0%B0%D0%B1%D0%BB%D0%BE%D0%BD_%D0%BF%D1%80%D0%BE%D0%B5%D0%BA%D1%82%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D1%8F)